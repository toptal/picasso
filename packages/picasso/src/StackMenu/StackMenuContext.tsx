import { ReactElement, createContext } from 'react'

export interface StackMenuContextProps {
  push?: (key: string, menu: ReactElement) => void
  pop?: () => void
  refresh?: (key: string, menu: ReactElement) => void
}

export default createContext<StackMenuContextProps>({})
