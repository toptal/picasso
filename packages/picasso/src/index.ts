export { BaseProps } from '@toptal/picasso-shared'
export type { ColorType, SpacingType, SizeType } from '@toptal/picasso-shared'
export { AccordionCompound as Accordion } from './AccordionCompound'
export type { AccordionProps } from './Accordion'
export { default as AccountSelect } from './AccountSelect'
export type { AccountSelectProps } from './AccountSelect'
export { default as Amount } from './Amount'
export type { AmountProps } from './Amount'
export { default as Autocomplete } from './Autocomplete'
export type {
  AutocompleteProps,
  Item as AutocompleteItem,
} from './Autocomplete'
export { AvatarCompound as Avatar } from './AvatarCompound'
export type { AvatarProps } from './Avatar'
export { ButtonCompound as Button } from './ButtonCompound'
export type { ButtonProps, VariantType as ButtonVariantType } from './Button'
export type { ButtonCheckboxProps } from './ButtonCheckbox'
export type { ButtonRadioProps } from './ButtonRadio'
export { CheckboxCompound as Checkbox } from './CheckboxCompound'
export type { CheckboxProps } from './Checkbox'
export type { CheckboxGroupProps } from './CheckboxGroup'
export { default as Container } from './Container'
export type { ContainerProps } from './Container'
export { DropdownCompound as Dropdown } from './DropdownCompound'
export type { DropdownProps } from './Dropdown'
export { default as DropdownArrow } from './DropdownArrow'
export type { DropdownArrowProps } from './DropdownArrow'
export { default as EnvironmentBanner } from './EnvironmentBanner'
export type { EnvironmentBannerProps } from './EnvironmentBanner'
export { default as FileInput } from './FileInput'
export type { FileInputProps } from './FileInput'
export { default as FileList } from './FileList'
export type { FileListProps } from './FileList'
export { FormCompound as Form } from './FormCompound'
export type { FormProps } from './Form'
export { default as FormHint } from './FormHint'
export type { FormHintProps } from './FormHint'
export { default as FormError } from './FormError'
export type { FormErrorProps } from './FormError'
export { default as FormLabel } from './FormLabel'
export type { FormLabelProps, RequiredDecoration } from './FormLabel'
export { GridCompound as Grid } from './GridCompound'
export type { GridProps, GridSize } from './Grid'
export { HelpboxCompound as Helpbox } from './HelpboxCompound'
export type { HelpboxProps } from './Helpbox'
export { default as Image } from './Image'
export type { ImageProps } from './Image'
export { default as Indicator } from './Indicator'
export type { IndicatorProps } from './Indicator'
export { default as Input } from './Input'
export type { InputProps } from './Input'
export { default as InputAdornment } from './InputAdornment'
export type { InputAdornmentProps } from './InputAdornment'
export { TagCompound as Tag } from './TagCompound'
export type { TagProps } from './Tag'
export type { TagRectangularProps } from './TagRectangular'
export { default as Link } from './Link'
export type { LinkProps } from './Link'
export { default as Loader } from './Loader'
export type { LoaderProps } from './Loader'
export { default as Logo } from './Logo'
export type { LogoProps } from './Logo'
export { MenuCompound as Menu } from './MenuCompound'
export type { MenuProps } from './Menu'
export { default as MenuItem } from './MenuItem'
export type { MenuItemProps } from './MenuItem'
export { ModalCompound as Modal } from './ModalCompound'
export type { ModalProps } from './Modal'
export { default as MonthSelect } from './MonthSelect'
export type { MonthSelectProps } from './MonthSelect'
export { NotificationCompound as Notification } from './NotificationCompound'
export type { NotificationProps } from './Notification'
export { default as NotificationActions } from './NotificationActions'
export type { NotificationActionsProps } from './NotificationActions'
export { default as NumberInput } from './NumberInput'
export type { NumberInputProps } from './NumberInput'
export { PageCompound as Page } from './PageCompound'
export type { PageProps } from './Page'
export { default as Pagination } from './Pagination'
export type { PaginationProps } from './Pagination'
export { default as Paper } from './Paper'
export type { PaperProps } from './Paper'
export { default as PromptModal } from './PromptModal'
export type { PromptModalProps } from './PromptModal'
export { RadioCompound as Radio } from './RadioCompound'
export type { RadioProps } from './Radio'
export type { RadioGroupProps } from './RadioGroup'
export { default as RichTextEditor } from './RichTextEditor'
export type {
  RichTextEditorProps,
  RichTextEditorChangeHandler,
} from './RichTextEditor'
export { default as Select } from './Select'
export type { SelectProps } from './Select'
export type {
  Option as SelectOption,
  ValueType as SelectValueType,
} from './SelectBase'
export type { PageSidebarProps } from './PageSidebar'
export { default as Slider } from './Slider'
export type { SliderProps } from './Slider'
export { default as ShowMore } from './ShowMore'
export type { ShowMoreProps } from './ShowMore'
export { StepperCompound as Stepper } from './StepperCompound'
export type { StepperProps } from './Stepper'
export type { StepperVerticalProps } from './StepperVertical'
export { TableCompound as Table } from './TableCompound'
export type { TableProps } from './Table'
export { default as TableFooter } from './TableFooter'
export type { TableFooterProps } from './TableFooter'
export { TabsCompound as Tabs } from './TabsCompound'
export type { TabsProps } from './Tabs'
export { TagSelectorCompound as TagSelector } from './TagSelectorCompound'
export type { TagSelectorProps } from './TagSelector'
export { default as Tooltip } from './Tooltip'
export type { TooltipProps } from './Tooltip'
export { default as Typography } from './Typography'
export type { TypographyProps } from './Typography'
export { default as UserBadge } from './UserBadge'
export type { UserBadgeProps } from './UserBadge'
export { default as YearSelect } from './YearSelect'
export type { YearSelectProps } from './YearSelect'
export { default as ApplicationUpdateNotification } from './ApplicationUpdateNotification'
export type { ApplicationUpdateNotificationProps } from './ApplicationUpdateNotification'
export { AlertCompound as Alert } from './AlertCompound'
export type { AlertProps } from './Alert'
export type { AlertInlineProps } from './AlertInline'
export { default as Switch } from './Switch'
export type { SwitchProps } from './Switch'
export { default as Quote } from './Quote'
export type { QuoteProps } from './Quote'
export { default as Rating } from './Rating'
export type { RatingStarsProps, RatingThumbsProps } from './Rating'
export {
  default as TreeView,
  TreeViewContainer,
  useTreeView,
  TreeNodeAvatar,
} from './TreeView'
export type { TreeNodeInterface } from './TreeView'
export { default as StaticTreeView } from './StaticTreeView'
export { default as TimePicker } from './TimePicker'
export type { TimePickerProps } from './TimePicker'
export { default as SkeletonLoader } from './SkeletonLoader'
export { default as PageHead } from './PageHead'
export type { PageHeadProps } from './PageHead'
export { default as ProgressBar } from './ProgressBar'
export type { ProgressBarProps } from './ProgressBar'
export { default as List } from './List'
export type { ListProps } from './List'
export { default as Badge } from './Badge'
export type { BadgeProps } from './Badge'
export { BreadcrumbsCompound as Breadcrumbs } from './BreadcrumbsCompound'
export type { BreadcrumbsProps } from './Breadcrumbs'
export { default as Calendar } from './Calendar'
export type { DateOrDateRangeType } from './Calendar'
export {
  default as DatePicker,
  datePickerParseDateString,
  DEFAULT_DATE_PICKER_DISPLAY_DATE_FORMAT,
  DEFAULT_DATE_PICKER_EDIT_DATE_FORMAT,
} from './DatePicker'
export type {
  DatePickerValue,
  DatePickerInputCustomValueParser,
  DatePickerInputCustomValueParserParameters,
  DatePickerProps,
} from './DatePicker'
export { default as Drawer } from './Drawer'
export type { DrawerProps } from './Drawer'
export { default as Dropzone, ErrorCode as DropzoneErrorCode } from './Dropzone'
export type {
  DropzoneProps,
  FileUpload as DropzoneFileUpload,
  FileError as DropzoneFileError,
  FileRejection as DropzoneFileRejection,
  DropEvent as DropzoneDropEvent,
} from './Dropzone'
export { OverviewBlockCompound as OverviewBlock } from './OverviewBlockCompound'
export type { OverviewBlockProps } from './OverviewBlock'
export { default as TypographyOverflow } from './TypographyOverflow'
export type { TypographyOverflowProps } from './TypographyOverflow'
export { default as EmptyState } from './EmptyState'
export { default as Section } from './Section'
export type { SectionProps } from './Section'
export { NoteCompound as Note } from './NoteCompound'
export type { NoteProps } from './Note'
export { default as Timeline } from './Timeline'
export type { TimelineProps } from './Timeline'
export { default as RichText } from './RichText'
export type { RichTextProps, ASTType } from './RichText'
export { default as PasswordInput } from './PasswordInput'
export type { PasswordInputProps } from './PasswordInput'
export { default as FieldRequirements } from './FieldRequirements'
export type {
  FieldRequirementsProps,
  FieldRequirement,
} from './FieldRequirements'
export type { Status as OutlinedInputStatus } from './OutlinedInput'
export { default as AvatarUpload } from './AvatarUpload'
export type {
  AvatarUploadProps,
  FileUpload as AvatarUploadFileUpload,
  FileError as AvatarUploadFileError,
  FileRejection as AvatarUploadFileRejection,
  DropEvent as AvatarUploadDropEvent,
} from './AvatarUpload'

// hygen code generator inserts export statements above this comment.
export * from './Icon'
