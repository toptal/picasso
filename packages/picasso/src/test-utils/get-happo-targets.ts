type Result = Record<string, any>

/**
 * Generates Happo targets for responsive visual tests
 *
 * @param screenWidths Screen widths to generate targets for
 * @returns {Result} Object with Happo targets
 */
export const getHappoTargets = (screenWidths: number[]) =>
  screenWidths.reduce<Result>((acc: Result, width: number) => {
    const name = `chrome-desktop-width-${width}`

    acc[name] = {
      name,
      browser: 'chrome',
      viewport: `${width}x1024`,
    }

    return acc
  }, {})
