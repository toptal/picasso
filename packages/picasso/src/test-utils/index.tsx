import type { ReactElement } from 'react'
import React from 'react'
import type { RenderOptions, RenderResult } from '@testing-library/react'
import { render } from '@testing-library/react'

import TestingPicasso from './TestingPicasso'

export type PicassoConfig = {
  titleCase: boolean
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  picassoConfig?: PicassoConfig
): RenderResult => {
  const { rerender, ...rest } = render(
    <TestingPicasso titleCase={picassoConfig?.titleCase}>{ui}</TestingPicasso>,
    options
  )

  const customRerender: typeof rerender = newUi =>
    rerender(
      <TestingPicasso titleCase={picassoConfig?.titleCase}>
        {newUi}
      </TestingPicasso>
    )

  return { rerender: customRerender, ...rest }
}

export * from '@testing-library/react'
export { customRender as render, TestingPicasso }
export { getCheckpoints } from './get-checkpoints'
