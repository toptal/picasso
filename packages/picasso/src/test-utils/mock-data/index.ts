export { default as tableMockData } from './table'
