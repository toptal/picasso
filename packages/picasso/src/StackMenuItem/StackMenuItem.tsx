import React, {
  forwardRef,
  useCallback,
  useContext,
  useEffect,
  useState
} from 'react'
import { OverridableComponent } from '@toptal/picasso-shared'

import MenuListItem, { MenuListItemProps } from '../MenuListItem'
import StackMenuContext, {
  StackMenuContextProps
} from '../StackMenu/StackMenuContext'

export type VariantType = 'light' | 'dark'

export type Props = Omit<MenuListItemProps, 'contentRef'>

const generateKey = (() => {
  let count = 0

  return () => String(++count)
})()

export const StackMenuItem: OverridableComponent<Props> = forwardRef<
  HTMLElement,
  Props
>(function StackMenuItem (props, ref) {
  const { className, style, menu, onClick, ...rest } = props
  const [key] = useState(generateKey)
  const { push, refresh } = useContext<StackMenuContextProps>(StackMenuContext)

  useEffect(() => {
    if (menu && refresh) {
      refresh(key, menu)
    }
  }, [key, menu, refresh])

  const handleClick = useCallback(
    (event: React.MouseEvent<HTMLElement, MouseEvent>) => {
      if (menu && push) {
        event.stopPropagation()
        push(key, menu)
      }

      if (onClick) {
        onClick(event)
      }
    },
    [key, menu, push, onClick]
  )

  return (
    <MenuListItem
      {...rest}
      ref={ref}
      className={className}
      style={style}
      menu={menu}
      onClick={handleClick}
    />
  )
})

StackMenuItem.displayName = 'StackMenuItem'

export default StackMenuItem
