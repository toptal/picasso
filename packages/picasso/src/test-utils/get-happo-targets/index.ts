import { getCheckpoints } from './get-checkpoints'
import { getHappoTargets } from './get-happo-targets'

const checkpoints = getCheckpoints()

export { getHappoTargets }

export const HAPPO_TARGETS = getHappoTargets(checkpoints)
