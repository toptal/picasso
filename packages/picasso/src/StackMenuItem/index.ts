import { OmitInternalProps } from '@toptal/picasso-shared'

import { Props } from './StackMenuItem'

export { default } from './StackMenuItem'
export type StackMenuItemProps = OmitInternalProps<Props>
