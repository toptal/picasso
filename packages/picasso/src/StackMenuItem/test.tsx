import React from 'react'
import { fireEvent, render } from '@toptal/picasso/test-utils'

import StackMenuItem from '../StackMenuItem'
import StackMenuContext from '../StackMenu/StackMenuContext'

describe('StackMenuItem', () => {
  it('renders', () => {
    const { container } = render(<StackMenuItem>Item</StackMenuItem>)

    expect(container).toMatchSnapshot()
  })

  it('delegates menu element', () => {
    const menu = <div />
    const refresh = jest.fn()

    render(
      <StackMenuContext.Provider value={{ refresh }}>
        <StackMenuItem menu={menu}>Item</StackMenuItem>
      </StackMenuContext.Provider>
    )

    expect(refresh).toHaveBeenCalledWith(expect.any(String), menu)
  })

  it('delegates clicks for toggling a menu', () => {
    const menu = <div />
    const push = jest.fn()

    const { getByTestId } = render(
      <StackMenuContext.Provider value={{ push }}>
        <StackMenuItem menu={menu} data-testid='item'>
          Item
        </StackMenuItem>
      </StackMenuContext.Provider>
    )

    fireEvent.click(getByTestId('item'))

    expect(push).toHaveBeenCalledWith(expect.any(String), menu)
  })
})
