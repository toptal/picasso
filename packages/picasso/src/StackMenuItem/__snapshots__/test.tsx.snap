// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackMenuItem renders 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <li
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root PicassoMenuListItem-root PicassoMenuListItem-light MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
      role="menuitem"
      tabindex="-1"
    >
      <div
        class="PicassoContainer-flex PicassoContainer-column PicassoMenuListItem-content"
      >
        <div
          class="PicassoContainer-centerAlignItems PicassoContainer-flex"
        >
          <span
            class="PicassoMenuListItem-stringContent"
          >
            Item
          </span>
        </div>
      </div>
    </li>
  </div>
</div>
`;
